### ISSUE TITLE
XLA spend too much time for compile graph

### ISSUE THREAD

**Issue by tiena2cva** [pytorch/xla#3125]
## ❓ Questions and Help

I want to fine tune wav2vec model on colab TPU and use [huggingface library](https://github.com/huggingface/transformers/tree/master/src/transformers/models/wav2vec2), which [support XLA](https://huggingface.co/blog/pytorch-xla)

But when I use TPU, training time got x100 slower than GPU. I have read [TROUBLESHOOTING doc](https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md) and here is my dataset:
- I use in memory dataset
- I use huggingface dataset library, a wrapper of pytorch dataset
- train dataset
  ```
  Dataset({
      features: ['input_values', 'labels'],
      num_rows: 4499
  })
  ```
- all data are torch tensor and have same size:
  - input_values: `torch.Size([80000])`
  - labels: `torch.Size([200])`

Do someone have experience about this case?
 
Here is metric debug information:
```
Metric: CompileTime
  TotalSamples: 159
  Accumulator: 02h18m48s875ms437.223us
  ValueRate: 934ms194.825us / second
  Rate: 0.0179656 / second
  Percentiles: 1%=009ms444.493us; 5%=013ms944.515us; 10%=017ms506.443us; 20%=111ms573.400us; 50%=40s497ms025.717us; 80%=02m45s961ms078.540us; 90%=02m52s515ms030.733us; 95%=02m55s692ms556.248us; 99%=02m59s031ms482.747us
Metric: DeviceLockWait
  TotalSamples: 7992
  Accumulator: 01m04s269ms519.738us
  ValueRate: 281.032us / second
  Rate: 8.7141 / second
  Percentiles: 1%=000.210us; 5%=000.742us; 10%=000.900us; 20%=001.081us; 50%=001.297us; 80%=001.861us; 90%=002.509us; 95%=004.090us; 99%=001ms203.708us
Metric: ExecuteTime
  TotalSamples: 6767
  Accumulator: 03m38s990ms293.374us
  ValueRate: 098ms555.513us / second
  Rate: 8.44178 / second
  Percentiles: 1%=002ms731.824us; 5%=002ms968.391us; 10%=002ms076.049us; 20%=002ms220.236us; 50%=003ms515.543us; 80%=003ms985.318us; 90%=004ms735.174us; 95%=113ms899.216us; 99%=114ms504.058us
Metric: InboundData
  TotalSamples: 6713
  Accumulator: 153.97MB
  ValueRate: 166.22KB / second
  Rate: 8.45022 / second
  Percentiles: 1%=1.00B; 5%=16.00B; 10%=24.00B; 20%=64.00B; 50%=64.00B; 80%=1.95KB; 90%=15.56KB; 95%=233.44KB; 99%=233.44KB
Metric: InputOutputAliasCount
  TotalSamples: 153
  Accumulator: 78190.00
  ValueRate: 8.83 / second
  Rate: 0.0172832 / second
  Percentiles: 1%=1.00; 5%=1.00; 10%=1.00; 20%=1.00; 50%=3.00; 80%=1586.00; 90%=1586.00; 95%=1586.00; 99%=1586.00
Metric: IrValueTensorToXlaData
  TotalSamples: 6176
  Accumulator: 38s519ms171.444us
  ValueRate: 017ms166.710us / second
  Rate: 8.06979 / second
  Percentiles: 1%=001ms014.776us; 5%=001ms110.834us; 10%=001ms175.317us; 20%=001ms273.447us; 50%=001ms486.577us; 80%=002ms806.875us; 90%=002ms028.233us; 95%=002ms276.481us; 99%=028ms177.526us
Metric: OutboundData
  TotalSamples: 6413
  Accumulator: 4.75GB
  ValueRate: 2.52MB / second
  Rate: 8.08313 / second
  Percentiles: 1%=0.00B; 5%=12.00B; 10%=16.00B; 20%=32.00B; 50%=64.00B; 80%=64.00B; 90%=64.00B; 95%=15.56KB; 99%=24.76MB
Metric: ReleaseDataHandlesTime
  TotalSamples: 11671
  Accumulator: 14s091ms749.430us
  ValueRate: 011ms143.770us / second
  Rate: 8.94147 / second
  Percentiles: 1%=489.011us; 5%=570.732us; 10%=621.612us; 20%=690.899us; 50%=835.321us; 80%=979.403us; 90%=001ms066.917us; 95%=001ms149.910us; 99%=027ms967.132us
Metric: TensorsGraphSize
  TotalSamples: 6767
  Accumulator: 3360478.00
  ValueRate: 2622.02 / second
  Rate: 8.44186 / second
  Percentiles: 1%=3.00; 5%=4.00; 10%=4.00; 20%=6.00; 50%=9.00; 80%=9.00; 90%=29.00; 95%=3869.00; 99%=3871.00
Metric: TransferFromServerTime
  TotalSamples: 6713
  Accumulator: 14s776ms842.830us
  ValueRate: 018ms274.338us / second
  Rate: 8.45022 / second
  Percentiles: 1%=001ms006.256us; 5%=001ms176.374us; 10%=001ms273.714us; 20%=001ms383.671us; 50%=002ms671.092us; 80%=002ms952.105us; 90%=002ms411.861us; 95%=004ms518.413us; 99%=026ms401.389us
Metric: TransferToServerTime
  TotalSamples: 6413
  Accumulator: 01m09s018ms524.202us
  ValueRate: 043ms278.963us / second
  Rate: 8.0831 / second
  Percentiles: 1%=001ms009.325us; 5%=001ms105.303us; 10%=001ms172.664us; 20%=001ms268.861us; 50%=001ms482.938us; 80%=002ms804.380us; 90%=002ms037.549us; 95%=002ms347.970us; 99%=272ms535.100us
Metric: TransferToServerTransformTime
  TotalSamples: 6413
  Accumulator: 02s709ms958.008us
  ValueRate: 001ms237.694us / second
  Rate: 8.08313 / second
  Percentiles: 1%=073.013us; 5%=085.822us; 10%=093.318us; 20%=100.597us; 50%=120.036us; 80%=182.507us; 90%=203.660us; 95%=226.156us; 99%=001ms343.332us
Counter: CachedCompile
  Value: 6608
Counter: CreateCompileHandles
  Value: 159
Counter: CreateDataHandles
  Value: 97441
Counter: CreateXlaTensor
  Value: 783076
Counter: DestroyDataHandles
  Value: 94053
Counter: DestroyXlaTensor
  Value: 779806
Counter: DeviceDataCacheMiss
  Value: 118
Counter: MarkStep
  Value: 259
Counter: ReleaseDataHandles
  Value: 94053
Counter: UncachedCompile
  Value: 159
Counter: XRTAllocateFromTensor_Empty
  Value: 83
Counter: XrtCompile_Empty
  Value: 144
Counter: XrtExecuteChained_Empty
  Value: 144
Counter: XrtExecute_Empty
  Value: 144
Counter: XrtMemoryInfo_Empty
  Value: 144
Counter: XrtRead_Empty
  Value: 144
Counter: XrtReleaseAllocationHandle_Empty
  Value: 144
Counter: XrtReleaseCompileHandle_Empty
  Value: 144
Counter: XrtSessionCount
  Value: 11
Counter: XrtSubTuple_Empty
  Value: 144
Counter: aten::_ctc_loss
  Value: 256
Counter: aten::_ctc_loss_backward
  Value: 52
Counter: aten::_local_scalar_dense
  Value: 261
Counter: aten::cumsum
  Value: 256
Counter: aten::floor_divide
  Value: 3584
Counter: aten::masked_select
  Value: 256
Counter: aten::multinomial
  Value: 52
Counter: aten::nonzero
  Value: 464
Counter: xla::_copy_from
  Value: 2798
Counter: xla::_log_softmax
  Value: 256
Counter: xla::_log_softmax_backward_data
  Value: 52
Counter: xla::_s_where
  Value: 104
Counter: xla::_softmax
  Value: 7124
Counter: xla::_softmax_backward_data
  Value: 1114
Counter: xla::_unsafe_view
  Value: 43256
Counter: xla::add
  Value: 29876
Counter: xla::add_
  Value: 102020
Counter: xla::addcdiv_
  Value: 20528
Counter: xla::addcmul
  Value: 16552
Counter: xla::addcmul_
  Value: 20528
Counter: xla::arange_out
  Value: 308
Counter: xla::as_strided
  Value: 1618
Counter: xla::bernoulli_
  Value: 6736
Counter: xla::bitwise_not_out
  Value: 256
Counter: xla::bmm
  Value: 18704
Counter: xla::clamp_min
  Value: 256
Counter: xla::clone
  Value: 156
Counter: xla::convolution_backward_overrideable
  Value: 52
Counter: xla::convolution_overrideable
  Value: 2048
Counter: xla::cumsum
  Value: 256
Counter: xla::div
  Value: 876
Counter: xla::div_
  Value: 6736
Counter: xla::empty
  Value: 12250
Counter: xla::empty_strided
  Value: 1618
Counter: xla::eq
  Value: 52
Counter: xla::expand
  Value: 568
Counter: xla::fill_
  Value: 668
Counter: xla::flip
  Value: 512
Counter: xla::ge
  Value: 512
Counter: xla::gelu
  Value: 9172
Counter: xla::gelu_backward
  Value: 1166
Counter: xla::index
  Value: 52
Counter: xla::index_put_
  Value: 668
Counter: xla::lt
  Value: 52
Counter: xla::masked_fill_
  Value: 52
Counter: xla::masked_select
  Value: 256
Counter: xla::max
  Value: 256
Counter: xla::mean
  Value: 460
Counter: xla::mm
  Value: 56832
Counter: xla::mul
  Value: 28312
Counter: xla::mul_
  Value: 61584
Counter: xla::native_batch_norm
  Value: 16552
Counter: xla::native_batch_norm_backward
  Value: 2280
Counter: xla::neg
  Value: 52
Counter: xla::nonzero
  Value: 464
Counter: xla::norm
  Value: 20836
Counter: xla::repeat
  Value: 204
Counter: xla::rsub
  Value: 256
Counter: xla::scatter_
  Value: 52
Counter: xla::select
  Value: 928
Counter: xla::slice
  Value: 1744
Counter: xla::sqrt
  Value: 20528
Counter: xla::squeeze
  Value: 2152
Counter: xla::stack
  Value: 52
Counter: xla::sub
  Value: 3840
Counter: xla::sub_
  Value: 5
Counter: xla::sum
  Value: 12324
Counter: xla::t
  Value: 63620
Counter: xla::transpose
  Value: 50410
Counter: xla::unsqueeze
  Value: 5584
Counter: xla::view
  Value: 265460
Counter: xla::zero_
  Value: 22680
Metric: XrtAllocateFromTensor
  TotalSamples: 7150
  Accumulator: 09s916ms589.597us
  Mean: 001ms066.349us
  StdDev: 003ms537.052us
  Rate: 8.25017 / second
  Percentiles: 25%=294.959us; 50%=367.649us; 80%=554.641us; 90%=762.044us; 95%=006ms647.786us; 99%=013ms343.839us
Metric: XrtCompaction
  TotalSamples: 1
  Accumulator: 05s050ms068.227us
  Mean: 05s050ms068.227us
  StdDev: 000.000us
  Percentiles: 25%=05s050ms068.227us; 50%=05s050ms068.227us; 80%=05s050ms068.227us; 90%=05s050ms068.227us; 95%=05s050ms068.227us; 99%=05s050ms068.227us
Metric: XrtCompile
  TotalSamples: 159
  Accumulator: 02h18m35s395ms299.430us
  Mean: 52s921ms725.154us
  StdDev: 40s321ms853.281us
  Rate: 0.0179655 / second
  Percentiles: 25%=35s876ms146.687us; 50%=40s460ms069.546us; 80%=02m45s799ms523.231us; 90%=02m51s342ms022.338us; 95%=02m55s500ms030.889us; 99%=02m59s826ms991.531us
Metric: XrtExecute
  TotalSamples: 6767
  Accumulator: 02m27s759ms204.977us
  Mean: 010ms831.227us
  StdDev: 029ms424.130us
  Rate: 8.44175 / second
  Percentiles: 25%=001ms094.031us; 50%=001ms251.192us; 80%=002ms525.844us; 90%=002ms973.935us; 95%=111ms253.261us; 99%=112ms740.193us
Metric: XrtExecutorEvict
  TotalSamples: 281
  Accumulator: 28.98GB
  Mean: 105.62MB
  StdDev: 152.13MB
  Rate: 0.035406 / second
  Percentiles: 25%=24.00KB; 50%=76.94MB; 80%=91.38MB; 90%=409.61MB; 95%=431.83MB; 99%=451.93MB
Metric: XrtReadLiteral
  TotalSamples: 7073
  Accumulator: 04s855ms633.759us
  Mean: 559.132us
  StdDev: 172.429us
  Rate: 8.50055 / second
  Percentiles: 25%=452.645us; 50%=537.386us; 80%=654.752us; 90%=795.957us; 95%=893.993us; 99%=001ms105.337us
Metric: XrtReleaseAllocation
  TotalSamples: 11671
  Accumulator: 01s023ms068.590us
  Mean: 031.303us
  StdDev: 019.679us
  Rate: 8.94145 / second
  Percentiles: 25%=020.817us; 50%=028.690us; 80%=037.099us; 90%=046.504us; 95%=057.288us; 99%=115.259us
Metric: XrtTryFreeMemory
  TotalSamples: 12
  Accumulator: 05s158ms199.948us
  Mean: 430ms849.996us
  StdDev: 294ms612.974us
  Rate: 0.00228191 / second
  Percentiles: 25%=262ms756.111us; 50%=412ms512.802us; 80%=649ms104.881us; 90%=713ms206.478us; 95%=997ms163.024us; 99%=997ms163.024us
```

**Comment by JackCaoG** [comment]
It is a known problem, ctc_loss is not lowered for pt/xla which will always fallback to CPU and caused a great slow down.  You can check https://github.com/pytorch/xla/issues/2681#issuecomment-894468034 for a bit more detail. We are still a bit tight on the resource side and don't have timeline for this lowering yet.

**Comment by tiena2cva** [comment]
> It is a known problem, ctc_loss is not lowered for pt/xla which will always fallback to CPU and caused a great slow down. You can check [#2681 (comment)](https://github.com/pytorch/xla/issues/2681#issuecomment-894468034) for a bit more detail. We are still a bit tight on the resource side and don't have timeline for this lowering yet.

Thank you so much. I am back to GPU now :((
